src/frontend/src/optimizer/plan_correlated_id_finder.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use crate::expr::{CorrelatedId, CorrelatedInputRef, ExprVisitor};
+use crate::optimizer::plan_node::{LogicalFilter, LogicalJoin, LogicalProject, PlanTreeNode};
+use crate::optimizer::plan_visitor::PlanVisitor;
+
+pub struct PlanCorrelatedIdFinder {
+    correlated_id_set: HashSet<CorrelatedId>,
+}
+
+impl PlanCorrelatedIdFinder {
+    pub fn new() -> Self {
+        Self {
+            correlated_id_set: HashSet::new(),
+        }
+    }
+
+    pub fn contains(&self, correlated_id: &CorrelatedId) -> bool {
+        self.correlated_id_set.contains(correlated_id)
+    }
+}
+
+impl PlanVisitor<()> for PlanCorrelatedIdFinder {
+    /// common subquery is project, filter and join
+
+    fn visit_logical_join(&mut self, plan: &LogicalJoin) {
+        let mut finder = ExprCorrelatedIdFinder::new();
+        plan.on()
+            .conjunctions
+            .iter()
+            .for_each(|expr| finder.visit_expr(expr));
+        self.correlated_id_set.extend(finder.correlated_id_set);
+
+        let mut iter = plan.inputs().into_iter();
+        self.visit(iter.next().unwrap());
+        iter.for_each(|input| {
+            self.visit(input);
+        });
```

Comment:
Why don't you just use `plan.inputs().into_iter().for_each(|input| self.visit(input));`?

Reply:
I just keep the same style of PlanVisitor default implementation.

Reply:
Done